{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import datetime\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib.ticker as tick\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as pplt\n",
    "import matplotlib.cm as cm\n",
    "import matplotlib.lines as lines\n",
    "from matplotlib import rcParams as rcp\n",
    "import plotly.graph_objects as go"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Setting variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#gene_name1 = input('Gene name 1: ')\n",
    "#gene_name2 = input('Gene name 2 (if wanted, else press enter): ')\n",
    "\n",
    "gene_name1 = 'brca1'\n",
    "gene_name2 = 'brca2'\n",
    "\n",
    "date = str(datetime.date.today())\n",
    "\n",
    "# joint file of gene\n",
    "df1 = pd.read_csv(r'joint_files_out/' + gene_name1.lower() + '_joint_out.csv')\n",
    "dfs = [df1]\n",
    "gene_names = [gene_name1]\n",
    "if gene_name2 != '':\n",
    "    df2 = pd.read_csv(r'joint_files_out/' + gene_name2.lower() + '_joint_out.csv')\n",
    "    dfs.append(df2)\n",
    "    gene_names.append(gene_name2)\n",
    "\n",
    "# SIFT_VEP and PolyPhen_VEP combined str and float values -> create one column for each\n",
    "for df in dfs:\n",
    "    df[['SIFT_VEP_str', 'SIFT_VEP_num']] = df['SIFT_VEP'].str.split('(', expand = True)\n",
    "    df['SIFT_VEP_num'] = df['SIFT_VEP_num'].str.strip(')').astype(float)\n",
    "    df[['PolyPhen_VEP_str', 'PolyPhen_VEP_num']] = df['PolyPhen_VEP'].str.split('(', expand = True)\n",
    "    df['PolyPhen_VEP_num'] = df['PolyPhen_VEP_num'].str.strip(')').astype(float)\n",
    "    df['Consequence_VEP'] = df['Consequence_VEP'].str.replace('_', ' ')\n",
    "    df['Consequence_VEP'] = df['Consequence_VEP'].str.replace(',', ', ')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# coding sequences as df\n",
    "coding_df1 = df1.loc[df1['Protein_position'].notnull()].copy()\n",
    "coding_dfs = [coding_df1]\n",
    "# unique positions (+ introns)\n",
    "all_pos1 = list(df1['POS'].unique())\n",
    "all_pos = [all_pos1]\n",
    "\n",
    "if gene_name2 != '':\n",
    "    coding_df2 = df2.loc[df2['Protein_position'].notnull()].copy()\n",
    "    coding_dfs.append(coding_df2)\n",
    "    all_pos2 = list(df2['POS'].unique())\n",
    "    all_pos.append(all_pos2)\n",
    "\n",
    "revs = []\n",
    "for df, cod in zip(dfs, coding_dfs):\n",
    "    if df.iloc[0]['Strand'] == '+':\n",
    "        cod.sort_values(by = ['POS'], ascending = True, ignore_index = True, inplace = True)\n",
    "        revs.append(False)\n",
    "    elif df.iloc[0]['Strand'] == '-':\n",
    "        cod.sort_values(by = ['POS'], ascending = False, ignore_index = True, inplace = True)\n",
    "        revs.append(True)\n",
    "    cod['Protein_position'] = cod['Protein_position'].astype(int)\n",
    "    cod['POS'] = cod['POS'].astype(int)\n",
    "\n",
    "# unique coding positions list and length\n",
    "codpos1 = list(coding_df1['POS'].unique())\n",
    "codpos = [codpos1]\n",
    "cod1_len = len(codpos1)\n",
    "cod_len = [cod1_len]\n",
    "if gene_name2 != '':\n",
    "    codpos2 = list(coding_df2['POS'].unique())\n",
    "    codpos.append(codpos2)\n",
    "    cod2_len = len(codpos2)-1\n",
    "    cod_len.append(cod2_len)\n",
    "\n",
    "for pos,cpos,r in zip(all_pos, codpos, revs):\n",
    "    pos.sort(reverse = r)\n",
    "    cpos.sort(reverse = r)\n",
    "\n",
    "# aa length per gene\n",
    "aa1_len = len(list(coding_df1['Protein_position'].unique()))\n",
    "aa2_len = len(list(coding_df2['Protein_position'].unique()))\n",
    "aa_len = [aa1_len, aa2_len]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# search for exon start/stop\n",
    "tuple_exons = []\n",
    "for df, cpos in zip(dfs, codpos):\n",
    "    exons_beg = []\n",
    "    exons_end = []\n",
    "    tuple_exons_in = []\n",
    "    if df.iloc[0]['Strand'] == '+':\n",
    "        for el in cpos:\n",
    "            if el-1 not in cpos:\n",
    "                exons_beg.append(el)\n",
    "            elif el+1 not in cpos:\n",
    "                exons_end.append(el)\n",
    "            else:\n",
    "                continue\n",
    "    elif df.iloc[0]['Strand'] == '-':\n",
    "        for el in cpos:\n",
    "            if el+1 not in cpos:\n",
    "                exons_beg.append(el)\n",
    "            elif el-1 not in cpos:\n",
    "                exons_end.append(el)\n",
    "            else:\n",
    "                continue\n",
    "    else:\n",
    "        pass\n",
    "    for x, y in zip(exons_beg, exons_end):\n",
    "        tuple_exons_in.append((x,y))\n",
    "    tuple_exons.append(tuple_exons_in)\n",
    "\n",
    "tuple_exons1 = tuple_exons[0]\n",
    "if len(tuple_exons) > 1:\n",
    "    tuple_exons2 = tuple_exons[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# exon numbers position\n",
    "txt_pos1 = []\n",
    "for x in range(len(tuple_exons1)):\n",
    "    if (int(x)+1)%2 != 0:\n",
    "        txt_pos1.append(-0.4)\n",
    "    elif (int(x)+1)%2 == 0:\n",
    "        txt_pos1.append(0.4)\n",
    "txt_pos = [txt_pos1]\n",
    "if gene_name2 != '':\n",
    "    txt_pos2 = []\n",
    "    for x in range(len(tuple_exons2)):\n",
    "        if (int(x)+1)%2 != 0:\n",
    "            txt_pos2.append(-0.4)\n",
    "        elif (int(x)+1)%2 == 0:\n",
    "            txt_pos2.append(0.4)\n",
    "    txt_pos.append(txt_pos2)\n",
    "else:\n",
    "    pass"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Overview of gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# figure for exons and introns\n",
    "def overview_fig(gene):\n",
    "    ind = gene_names.index(str(gene))\n",
    "    fig,ax = plt.subplots(figsize=(20, 7))\n",
    "    plt.ylim([-1, 1])\n",
    "    plt.axis('off')\n",
    "    fig.suptitle('Gene model of ' + gene_names[ind].upper() + ' gene in reading direction, CDS = ' + \n",
    "                 str(cod_len[ind]/1000) + ' kb, protein = ' + str('{:,.0f}'.format(aa_len[ind])) + ' aa', \n",
    "                 fontsize = 25, fontweight = 'bold')\n",
    "    plt.xlim(tuple_exons[ind][0][0], tuple_exons[ind][-1][1])\n",
    "    if dfs[ind].iloc[0]['Strand'] == '+':\n",
    "        ax.add_patch(pplt.Rectangle((tuple_exons[ind][0][0], -.1), tuple_exons[ind][-1][1], .2, \n",
    "                                    fc = 'dodgerblue', ec = 'dodgerblue', alpha = 1, label = 'Introns'))\n",
    "        ax.arrow(tuple_exons[ind][0][0], -0.7, abs(tuple_exons[ind][-1][1]-tuple_exons[ind][0][0])-1000, 0, \n",
    "                 head_width = 0.05, head_length = 1000, fc = 'k', ec = 'k')\n",
    "        for tup,l in zip(tuple_exons[ind], range(len(tuple_exons[ind]))):\n",
    "            ax.add_patch(pplt.Rectangle((int(tup[0]), -.3), abs(int(tup[0])-int(tup[1])), .6, fc = 'orange', \n",
    "                                        ec = 'orange', alpha = 1))\n",
    "            plt.text(int(tup[0]-300), txt_pos[ind][int(l)], str((int(l)+2)), size = 22)\n",
    "        plt.text(tuple_exons[ind][-1][1]-23000, -0.6, 'End: CHROM ' + str(dfs[ind].iloc[0]['CHROM']) + \n",
    "                 ', POS ' + str('{:,.0f}'.format(tuple_exons[ind][-1][1])), size = 25)\n",
    "        plt.text((tuple_exons[ind][0][0] + abs(tuple_exons[ind][-1][1]-tuple_exons[ind][0][0])/2)-8000, -0.6, \n",
    "                 'Gene Length: ' + str(abs(tuple_exons[ind][-1][1]-tuple_exons[ind][0][0])/1000) + ' kb', \n",
    "                 size = 25)\n",
    "    elif dfs[ind].iloc[0]['Strand'] == '-':\n",
    "        ax.add_patch(pplt.Rectangle((tuple_exons[ind][-1][1], -.1), tuple_exons[ind][0][0], .2, \n",
    "                                    fc = 'dodgerblue', ec = 'dodgerblue', alpha = 1, label = 'Introns'))\n",
    "        ax.arrow(tuple_exons[ind][0][0], -0.7, -abs(tuple_exons[ind][-1][1]-tuple_exons[ind][0][0])+1000, 0, \n",
    "                 head_width = 0.05, head_length = 1000, fc = 'k', ec = 'k')\n",
    "        for tup,l in zip(tuple_exons[ind], range(len(tuple_exons[ind]))):\n",
    "            ax.add_patch(pplt.Rectangle((int(tup[1]), -.3), abs(int(tup[0])-int(tup[1])), .6, fc = 'orange', \n",
    "                                        ec = 'orange', alpha = 1))\n",
    "            plt.text(int(tup[1]+300), txt_pos[ind][int(l)], str((int(l)+2)), size = 22)\n",
    "        plt.text(tuple_exons[ind][-1][1]+23000, -0.6, 'End: CHROM ' + str(dfs[ind].iloc[0]['CHROM']) + \n",
    "                 ', POS ' + str('{:,.0f}'.format(tuple_exons[ind][-1][1])), size = 25)\n",
    "        plt.text((tuple_exons[ind][0][0] - abs(tuple_exons[ind][-1][1]-tuple_exons[ind][0][0])/2)+8000, -0.6, \n",
    "                 'Gene Length: ' + str(abs(tuple_exons[ind][-1][1]-tuple_exons[ind][0][0])/1000) + ' kb', \n",
    "                 size = 25)\n",
    "\n",
    "    plt.text(tuple_exons[ind][0][0], 0.73, 'Exons', c = 'orange', size = 25, fontweight = 'bold')\n",
    "    plt.text(tuple_exons[ind][0][0], 0.6, 'Introns', c = 'dodgerblue', size = 25, fontweight = 'bold')\n",
    "\n",
    "    plt.text(tuple_exons[ind][0][0], -0.9, ('Reading Direction, \"' + str(dfs[ind].iloc[0]['Strand'])+'\" Strand'+\n",
    "                                            ', Transcript ' + str(dfs[ind].iloc[0]['Ref_Trans_VEP'])), size = 25)\n",
    "\n",
    "    plt.text(tuple_exons[ind][0][0], -0.6, 'Start: CHROM ' + str(dfs[ind].iloc[0]['CHROM']) + ', POS ' + \n",
    "             str('{:,.0f}'.format(tuple_exons[ind][0][0])), size = 25)\n",
    "\n",
    "\n",
    "    plt.gcf().subplots_adjust(bottom = 0.05, top = 0.95, left = 0.01, right = 0.95);\n",
    "\n",
    "    plt.savefig(r'figures/' + date + '_' + gene_names[ind].lower() + '_exins_ov.pdf')\n",
    "    plt.savefig(r'figures/' + date + '_' + gene_names[ind].lower() + '_exins_ov.jpg')\n",
    "\n",
    "for gene in gene_names:\n",
    "    overview_fig(gene)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Overview of annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for pie chart with origin of variants, plus info of exon/intron\n",
    "sizes = []\n",
    "labels = []\n",
    "cols = []\n",
    "sizes_sg = []\n",
    "labels_sg = []\n",
    "cols_sg = []\n",
    "\n",
    "for df in dfs:\n",
    "    sizes.append([len(df[df['AC_gnomAD'].notnull() & df['ClinVar_gnomAD'].notnull()]), \n",
    "                  len(df[df['AC_gnomAD'].notnull() & df['ClinVar_gnomAD'].isnull()]), \n",
    "                  len(df[df['Cancer_Type_cBP'].notnull()]),\n",
    "                  len(df[df['Trans_Version_Flossies'].notnull()]), \n",
    "                  len(df[df['NCT_SAMPLE_NAME'].notnull()]), \n",
    "                 ])\n",
    "    labels.append(['GnomAD (+ ClinVar)',\n",
    "                    'GnomAD (w/o ClinVar)',\n",
    "                    'cBioPortal',\n",
    "                    'Flossies',\n",
    "                    'MASTER/NCT project',\n",
    "                  ])\n",
    "    cols.append(['red', 'orange', 'green', 'dodgerblue', 'magenta'])\n",
    "    sizes_sg.append([len(df[df['AC_gnomAD'].notnull() & df['ClinVar_gnomAD'].notnull() & df['EXON'].notnull()]), \n",
    "            len(df[df['AC_gnomAD'].notnull() & df['ClinVar_gnomAD'].notnull() & df['EXON'].isnull()]), \n",
    "            len(df[df['AC_gnomAD'].notnull() & df['ClinVar_gnomAD'].isnull() & df['EXON'].notnull()]), \n",
    "            len(df[df['AC_gnomAD'].notnull() & df['ClinVar_gnomAD'].isnull() & df['EXON'].isnull()]), \n",
    "            len(df[df['Cancer_Type_cBP'].notnull() & df['EXON'].notnull()]), \n",
    "            len(df[df['Cancer_Type_cBP'].notnull() & df['EXON'].isnull()]), \n",
    "            len(df[df['Trans_Version_Flossies'].notnull() & df['EXON'].notnull()]), \n",
    "            len(df[df['Trans_Version_Flossies'].notnull() & df['EXON'].isnull()]), \n",
    "            len(df[df['NCT_SAMPLE_NAME'].notnull() & df['EXON'].notnull()]),\n",
    "            len(df[df['NCT_SAMPLE_NAME'].notnull() & df['EXON'].isnull()]),\n",
    "           ])\n",
    "    labels_sg.append(['E', 'I', 'E', 'I', 'E', 'I', 'E', 'I', 'E', 'I'])\n",
    "    cols_sg.append(['lightcoral', 'lightsalmon', 'gold', 'bisque', 'chartreuse', 'palegreen', \n",
    "          'deepskyblue', 'lightskyblue', 'violet', 'lightpink'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "fig = plt.subplots(figsize=(10, 12))\n",
    "# pie plot in coding sequence + 75 bp padding for gene 1, exon vs. intron\n",
    "rcp.update({'hatch.color':'gray'})\n",
    "if gene_name2 != '':\n",
    "    ax1 = plt.subplot(2,1,1)\n",
    "else:\n",
    "    ax1 = plt.subplot(1,1,1)\n",
    "ax1.set_title('Origin of variants in ' + gene_name1.upper() + ' gene', fontsize = 18, fontweight = 'bold')\n",
    "ax1.pie(sizes[0], labels = labels[0], autopct = '%1.1f%%', startangle = 180, pctdistance = 0.85, \n",
    "        colors = cols[0], counterclock = False, radius = 1, wedgeprops = dict(width = 0.3), \n",
    "        textprops = {'fontsize':15})\n",
    "piech1 = ax1.pie(sizes_sg[0], autopct = '%1.1f%%', startangle = 180, pctdistance = 0.8, \n",
    "            colors = cols_sg[0], textprops = {'fontsize':15},\n",
    "            counterclock = False, radius = 0.7, wedgeprops = dict(width = 0.3))\n",
    "for i in range(len(piech1[0])):\n",
    "    if (i+1)%2 == 0:\n",
    "        piech1[2][i].set_color('mediumblue')\n",
    "    if (i+1)%2 != 0:\n",
    "        piech1[2][i].set_color('darkmagenta');\n",
    "ax1.legend(handles = [pplt.Patch(label = 'Exons', fill = False, edgecolor = 'darkmagenta'), \n",
    "                      pplt.Patch(label = 'Introns', fill = False, edgecolor = 'mediumblue')], \n",
    "            bbox_to_anchor = (0.7, 0.9), loc = 'lower left', ncol = 2, labelcolor = ['darkmagenta', 'mediumblue'], fontsize = 15)\n",
    "\n",
    "# pie plot in coding sequence + 75 bp padding for gene 2, exon vs. intron\n",
    "if gene_name2 != '':\n",
    "    ax2 = plt.subplot(2,1,2)\n",
    "    ax2.set_title('Origin of variants in ' + gene_name2.upper() + ' gene', fontsize = 18, fontweight = 'bold')\n",
    "    ax2.pie(sizes[1], labels = labels[1], autopct = '%1.1f%%', startangle = 180, pctdistance = 0.85, \n",
    "            colors = cols[1], counterclock = False, radius = 1, wedgeprops = dict(width = 0.3), \n",
    "            textprops = {'fontsize':15})\n",
    "    piech2 = ax2.pie(sizes_sg[1], autopct = '%1.1f%%', startangle = 180, pctdistance = 0.8, \n",
    "                colors = cols_sg[1], textprops = {'fontsize':15},\n",
    "                counterclock = False, radius = 0.7, wedgeprops = dict(width = 0.3))\n",
    "    for i in range(len(piech2[0])):\n",
    "        if (i+1)%2 == 0:\n",
    "            piech2[2][i].set_color('mediumblue')\n",
    "        if (i+1)%2 != 0:\n",
    "            piech2[2][i].set_color('darkmagenta');\n",
    "    plt.tight_layout()\n",
    "    plt.gcf().subplots_adjust(bottom = 0.03, top = 0.95, left = 0.01, right = 0.97)\n",
    "    plt.savefig(r'figures/' + date + '_' + gene_name1.lower() + '_' + gene_name2.lower() + '_orig_vars.pdf')\n",
    "    plt.savefig(r'figures/' + date + '_' + gene_name1.lower() + '_' + gene_name2.lower() + '_orig_vars.jpg')\n",
    "else:\n",
    "    plt.tight_layout()\n",
    "    plt.gcf().subplots_adjust(bottom = 0.03, top = 0.95, left = 0.01, right = 0.97)\n",
    "    plt.savefig(r'figures/' + date + '_' + gene_name1.lower() + '_orig_vars.pdf')\n",
    "    plt.savefig(r'figures/' + date + '_' + gene_name1.lower() + '_orig_vars.jpg')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Overview of ClinVar/VEP impact"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#synthetic variants (expected) vs. variants in databases (observed)\n",
    "all_vars = []\n",
    "syn_vars_in = []\n",
    "syn_vars_ex = []\n",
    "db_vars_in = []\n",
    "db_vars_ex = []\n",
    "# value counts as df for all groups\n",
    "for df in dfs:\n",
    "    all_vars.append(pd.DataFrame(df['Consequence_VEP'].value_counts()))\n",
    "    syn_vars_in.append(pd.DataFrame(df['Consequence_VEP'][(df['Trans_Version_Syn'].notnull())&\n",
    "                                             (df['INTRON'].notnull())].value_counts()))\n",
    "    syn_vars_ex.append(pd.DataFrame(df['Consequence_VEP'][(df['Trans_Version_Syn'].notnull())&\n",
    "                                             (df['EXON'].notnull())].value_counts()))\n",
    "    db_vars_in.append(pd.DataFrame(df['Consequence_VEP'][(df['Trans_Version_Flossies'].notnull())|\n",
    "                                            (df1['Cancer_Type_cBP'].notnull())|\n",
    "                                            (df1['AF_gnomAD'].notnull())|\n",
    "                                            (df1['NCT_SAMPLE_NAME'].notnull())&\n",
    "                                            (df['INTRON'].notnull())].value_counts()))\n",
    "    db_vars_ex.append(pd.DataFrame(df['Consequence_VEP'][(df['Trans_Version_Flossies'].notnull())|\n",
    "                                            (df1['Cancer_Type_cBP'].notnull())|\n",
    "                                            (df1['AF_gnomAD'].notnull())|\n",
    "                                            (df1['NCT_SAMPLE_NAME'].notnull())&\n",
    "                                            (df['INTRON'].notnull())].value_counts()))\n",
    "# index as column, append right suffixes\n",
    "jnt_vars = []\n",
    "sufx = ['_all', '_syn_in', '_syn_ex', '_db_in', '_db_ex']\n",
    "for i,el in enumerate([all_vars,syn_vars_in,syn_vars_ex,db_vars_in,db_vars_ex]):\n",
    "    for data in el:\n",
    "        data['index_1'] = data.index\n",
    "    jnt_vars.append(el[0].merge(el[1], on = 'index_1', suffixes = [(sufx[i]+'_1'), (sufx[i]+'_2')], \n",
    "                                how = 'outer'))\n",
    "# merge all together\n",
    "merged_vars = jnt_vars[0].merge(jnt_vars[1], on = 'index_1', how = 'outer').merge(jnt_vars[2], on = 'index_1',\n",
    "                            how = 'outer').merge(jnt_vars[3], on = 'index_1', how = 'outer').merge(jnt_vars[4],\n",
    "                            on = 'index_1', how = 'outer')\n",
    "inex_cols = []\n",
    "for x in merged_vars['index_1']:\n",
    "    if x in ['intron variant', 'splice region variant, intron variant', 'splice donor variant', \n",
    "             'splice acceptor variant']:\n",
    "        inex_cols.append('dodgerblue')\n",
    "    else:\n",
    "        inex_cols.append('orange')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# variant consequences: relative values\n",
    "ct_syn_all1 = merged_vars['Consequence_VEP_syn_in_1'].sum() + merged_vars['Consequence_VEP_syn_ex_1'].sum()\n",
    "ct_syn_all2 = merged_vars['Consequence_VEP_syn_in_2'].sum() + merged_vars['Consequence_VEP_syn_ex_2'].sum()\n",
    "ct_db_all1 = merged_vars['Consequence_VEP_db_in_1'].sum() + merged_vars['Consequence_VEP_db_ex_1'].sum()\n",
    "ct_db_all2 = merged_vars['Consequence_VEP_db_in_2'].sum() + merged_vars['Consequence_VEP_db_ex_2'].sum()\n",
    "SynToDb1 = ct_db_all1/ct_syn_all1\n",
    "SynToDb2 = ct_db_all2/ct_syn_all2\n",
    "\n",
    "for col1,col2 in zip(['Consequence_VEP_syn_in_1', 'Consequence_VEP_syn_ex_1'], \n",
    "                     ['Consequence_VEP_syn_in_2', 'Consequence_VEP_syn_ex_2']):\n",
    "    merged_vars[(col1 + '_rel')] = merged_vars[col1]*SynToDb1\n",
    "    merged_vars[(col2 + '_rel')] = merged_vars[col2]*SynToDb2\n",
    "\n",
    "G1ToG2 = ct_syn_all1/ct_syn_all2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for pathogenicity by ClinVar, all, exons, introns\n",
    "path1_val = df1['ClinVar_gnomAD'][df1['ClinVar_gnomAD'].notnull()].value_counts()\n",
    "path1_val_df = pd.DataFrame(path1_val)\n",
    "path1_val_ex1_df = pd.DataFrame(df1['ClinVar_gnomAD'][(df1['ClinVar_gnomAD'].notnull()) &\n",
    "                                      (df1['EXON'].notnull())].value_counts())\n",
    "path1_val_in1_df = pd.DataFrame(df1['ClinVar_gnomAD'][(df1['ClinVar_gnomAD'].notnull()) &\n",
    "                                      (df1['INTRON'].notnull())].value_counts())\n",
    "# for impact by VEP\n",
    "imp1 = df1['Consequence_VEP'][df1['Consequence_VEP'].notnull()].value_counts()\n",
    "\n",
    "# for gene 2\n",
    "if gene_name2 != '':\n",
    "    path2_val = df2['ClinVar_gnomAD'][df2['ClinVar_gnomAD'].notnull()].value_counts()\n",
    "    path2_val_df = pd.DataFrame(path2_val)\n",
    "    path2_val_ex2_df = pd.DataFrame(df2['ClinVar_gnomAD'][(df2['ClinVar_gnomAD'].notnull()) &\n",
    "                                          (df2['EXON'].notnull())].value_counts())\n",
    "    path2_val_in2_df = pd.DataFrame(df2['ClinVar_gnomAD'][(df2['ClinVar_gnomAD'].notnull()) &\n",
    "                                          (df2['INTRON'].notnull())].value_counts())\n",
    "    imp2 = df2['Consequence_VEP'][df2['Consequence_VEP'].notnull()].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# pathogenicity, ClinVar order\n",
    "for el in [path1_val_df, path1_val_ex1_df, path1_val_in1_df, path2_val_df, path2_val_ex2_df, \n",
    "           path2_val_in2_df]:\n",
    "    el['index_1'] = el.index\n",
    "all_pat1_df = path1_val_df.merge(path1_val_ex1_df, on = 'index_1', suffixes = ['_all1', '_ex1']\n",
    "                  ).merge(path1_val_in1_df, on = 'index_1')\n",
    "all_pat1_df = all_pat1_df.rename(columns = {'ClinVar_gnomAD':'ClinVar_gnomAD_in1'})\n",
    "all_pat2_df = path2_val_df.merge(path2_val_ex2_df, on = 'index_1', suffixes = ['_all2', '_ex2']\n",
    "                  ).merge(path2_val_in2_df, on = 'index_1')\n",
    "all_pat2_df = all_pat2_df.rename(columns = {'ClinVar_gnomAD':'ClinVar_gnomAD_in2'})\n",
    "all_pat_df = all_pat1_df.merge(all_pat2_df, on = 'index_1', how = 'outer')\n",
    "# --> with these stacked bar plot\n",
    "clinvar_order = pd.CategoricalDtype(['Pathogenic', 'Pathogenic/Likely pathogenic', \n",
    "                                     'Likely pathogenic',\n",
    "                                     'Conflicting interpretations of pathogenicity',\n",
    "                                     'Likely benign', 'Benign/Likely benign', 'Benign',\n",
    "                                     'Uncertain significance', 'not provided'], ordered = True)\n",
    "all_pat_df['index_1'] = all_pat_df['index_1'].astype(clinvar_order)\n",
    "all_pat_df = all_pat_df.sort_values('index_1')\n",
    "\n",
    "for el in ['ClinVar_gnomAD_all2', 'ClinVar_gnomAD_ex2', 'ClinVar_gnomAD_in2']:\n",
    "    all_pat_df[el+'_rel'] = all_pat_df[el]*G1ToG2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure()\n",
    "axes = plt.subplots(figsize=(12, 12))\n",
    "ax1 = plt.subplot(2,2,1)\n",
    "ax2 = plt.subplot(2,2,3, sharey = ax1, sharex = ax1)\n",
    "ax3 = plt.subplot(2,2,2)\n",
    "ax4 = plt.subplot(2,2,4, sharey = ax3, sharex = ax3)\n",
    "\n",
    "or_leg = pplt.Patch(color = 'orange', label = 'in exons')\n",
    "blue_leg = pplt.Patch(color = 'dodgerblue', label = 'in introns')\n",
    "g1_leg = pplt.Patch(color = 'black', label = 'expected variants', fill = False)\n",
    "g2_leg = pplt.Patch(color = 'black', hatch = 'x', label = 'observed variants', fill = False)\n",
    "\n",
    "# impact of variants gene 1 (expected vs. observed)\n",
    "ar1 = np.arange(len(merged_vars['index_1']))\n",
    "wid1 = 0.35\n",
    "ax1.set_xticks(ar1)\n",
    "#ax1.set_xticklabels(merged_vars['index_1'], rotation = 45, ha = 'right')\n",
    "ax1.set_title('Impact of variants in ' + gene_name1.upper() + \n",
    "              ' gene', fontsize = 14, fontweight = 'bold')\n",
    "bar1 = ax1.bar(ar1-wid1/2, merged_vars['Consequence_VEP_syn_in_1_rel'], color = inex_cols, width = wid1, \n",
    "               label = gene_name1.upper(), edgecolor = 'black')\n",
    "bar2 = ax1.bar(ar1-wid1/2, merged_vars['Consequence_VEP_syn_ex_1_rel'], color = inex_cols, width = wid1, \n",
    "               label = gene_name1.upper(), edgecolor = 'black')\n",
    "bar3 = ax1.bar(ar1+wid1/2, merged_vars['Consequence_VEP_db_in_1'], color = inex_cols, width = wid1, \n",
    "               label = gene_name1.upper(), hatch = 'x', edgecolor = 'black')\n",
    "bar4 = ax1.bar(ar1+wid1/2, merged_vars['Consequence_VEP_db_ex_1'], color = inex_cols, width = wid1, \n",
    "               label = gene_name1.upper(), hatch = 'x', edgecolor = 'black')\n",
    "ax1.get_xaxis().set_visible(False)\n",
    "ax1.set_ylabel('Variant count (log scale)', fontsize = 13)\n",
    "ax1.set_yscale('log')\n",
    "\n",
    "# impact of variants gene 2 (expected vs. observed)\n",
    "ar2 = np.arange(len(merged_vars['index_1']))\n",
    "wid2 = 0.35\n",
    "ax2.set_xticks(ar2)\n",
    "ax2.set_xticklabels(merged_vars['index_1'], rotation = 45, ha = 'right')\n",
    "ax2.set_title('Impact of variants in ' + gene_name2.upper() + \n",
    "              ' gene (rel. to ' + gene_name1.upper() + ')', fontsize = 14, fontweight = 'bold')\n",
    "bar21 = ax2.bar(ar2-wid2/2, merged_vars['Consequence_VEP_syn_in_2_rel'], color = inex_cols, width = wid2, \n",
    "               label = gene_name2.upper(), edgecolor = 'black')\n",
    "bar22 = ax2.bar(ar2-wid2/2, merged_vars['Consequence_VEP_syn_ex_2_rel'], color = inex_cols, width = wid2, \n",
    "               label = gene_name2.upper(), edgecolor = 'black')\n",
    "bar23 = ax2.bar(ar2+wid2/2, merged_vars['Consequence_VEP_db_in_2'], color = inex_cols, width = wid2, \n",
    "               label = gene_name2.upper(), hatch = 'x', edgecolor = 'black')\n",
    "bar24 = ax2.bar(ar2+wid2/2, merged_vars['Consequence_VEP_db_ex_2'], color = inex_cols, width = wid2, \n",
    "               label = gene_name2.upper(), hatch = 'x', edgecolor = 'black')\n",
    "ax2.set_ylabel('Variant count (log scale)', fontsize = 13)\n",
    "ax2.set_yscale('log')\n",
    "plt.tight_layout()\n",
    "#'''\n",
    "ar3 = np.arange(len(all_pat_df['index_1']))\n",
    "#wid3 = 0.35\n",
    "ax3.set_xticks(ar3)\n",
    "ax3.set_xticklabels(all_pat_df['index_1'], rotation = 45, ha = 'right')\n",
    "ax3.set_title(('ClinVar annotations in ' + gene_name1.upper() \n",
    "               + ' gene'), fontsize = 14, fontweight = 'bold')\n",
    "ax3.bar(ar3, all_pat_df['ClinVar_gnomAD_ex1'],\n",
    "       label = 'exons', color = 'orange', edgecolor = 'black')\n",
    "ax3.bar(ar3, all_pat_df['ClinVar_gnomAD_in1'],\n",
    "        label = 'introns', bottom = all_pat_df['ClinVar_gnomAD_ex1'], \n",
    "        color = 'dodgerblue', edgecolor = 'black')\n",
    "ax3.get_xaxis().set_visible(False)\n",
    "#ax3.set_yscale('log')\n",
    "ax3.set_ylabel('Variant count', fontsize = 13)\n",
    "#'''\n",
    "#'''\n",
    "ar4 = np.arange(len(all_pat_df['index_1']))\n",
    "ax4.set_xticks(ar3)\n",
    "ax4.set_xticklabels(all_pat_df['index_1'], rotation = 45, ha = 'right')\n",
    "ax4.set_title(('ClinVar annotations in ' + gene_name2.upper() \n",
    "               + ' gene (rel. to ' + gene_name1.upper() + ')'), fontsize = 14, fontweight = 'bold')\n",
    "ax4.bar(ar4, all_pat_df['ClinVar_gnomAD_ex2_rel'],\n",
    "       label = 'exons', color = 'orange', edgecolor = 'black')\n",
    "ax4.bar(ar4, all_pat_df['ClinVar_gnomAD_in2_rel'],\n",
    "        label = 'introns', bottom = all_pat_df['ClinVar_gnomAD_ex2_rel'], \n",
    "        color = 'dodgerblue', edgecolor = 'black')\n",
    "#ax3.get_yaxis().set_visible(False)\n",
    "#ax4.set_yscale('log')\n",
    "ax4.set_ylabel('Variant count', fontsize = 13)\n",
    "\n",
    "ax2.legend(handles = [g1_leg, g2_leg, blue_leg, or_leg], bbox_to_anchor = (-0.2, 1.08), loc = 3, ncol = 4)\n",
    "#'''\n",
    "plt.tight_layout()\n",
    "plt.gcf().subplots_adjust(bottom = 0.25, top = 0.95, left = 0.08, right = 0.97)\n",
    "\n",
    "plt.savefig(r'figures/' + date + '_' + gene_name1.lower() + '_' + gene_name2.lower() +\n",
    "            '_vars_cons_impact.pdf')\n",
    "plt.savefig(r'figures/' + date + '_' + gene_name1.lower() + '_' + gene_name2.lower() +\n",
    "            '_vars_cons_impact.jpg')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Consequences of pathogenic variants (, ClinVar of most common variants)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# consequences of pathogenic variants\n",
    "cons_pat1_df = pd.DataFrame(df1['Consequence_VEP'][df1['ClinVar_gnomAD'] == 'Pathogenic'].value_counts())\n",
    "# clinvar annotations of most common consequence\n",
    "clinv_miss1_df = pd.DataFrame(df1['ClinVar_gnomAD'][df1['Consequence_VEP'] == \n",
    "                                                    'missense variant'].value_counts())\n",
    "\n",
    "if gene_name2 != '':\n",
    "    cons_pat2_df = pd.DataFrame(df2['Consequence_VEP'][df2['ClinVar_gnomAD'] == 'Pathogenic'].value_counts())\n",
    "    clinv_miss2_df = pd.DataFrame(df2['ClinVar_gnomAD'][df2['Consequence_VEP'] \n",
    "                                                        == 'missense variant'].value_counts())\n",
    "# index, merge\n",
    "for el in [cons_pat1_df, cons_pat2_df, clinv_miss1_df, clinv_miss2_df]:\n",
    "    el['index_1'] = el.index\n",
    "cons_patall_df = cons_pat1_df.merge(cons_pat2_df, on = 'index_1', how = 'outer', \n",
    "                                    suffixes = ('_1', '_2'))\n",
    "cons_patall_df['Consequence_VEP_2_rel'] = cons_patall_df['Consequence_VEP_2']*G1ToG2\n",
    "\n",
    "clinv_missall_df = clinv_miss1_df.merge(clinv_miss2_df, on = 'index_1', how = 'outer', \n",
    "                                        suffixes = ('_1', '_2'))\n",
    "clinv_missall_df['index_1'] = clinv_missall_df['index_1'].astype(clinvar_order)\n",
    "clinv_missall_df = clinv_missall_df.sort_values('index_1')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.subplots(figsize=(11, 6))\n",
    "# consequences of pathogenic variants in gene 1,2\n",
    "ar4 = np.arange(len(cons_patall_df['index_1']))\n",
    "wid4 = 0.35\n",
    "ax4 = plt.subplot(1,1,1)\n",
    "ax4.set_xticks(ar4)\n",
    "ax4.set_xticklabels(cons_patall_df['index_1'], rotation = 45, ha = 'right', fontsize = 13)\n",
    "\n",
    "ax4.set_title(('Consequences of pathogenic variants in ' + gene_name1.upper() + ' and ' + \n",
    "               gene_name2.upper() + ' gene (rel. to ' + gene_name1.upper() + ')'), \n",
    "              fontsize = 15, fontweight = 'bold')\n",
    "\n",
    "ax4.bar(ar4-wid4/2, cons_patall_df['Consequence_VEP_1'], width = wid4, color = 'white', \n",
    "        edgecolor = 'black', label = gene_name1.upper())\n",
    "ax4.bar(ar4+wid4/2, cons_patall_df['Consequence_VEP_2_rel'], width = wid4, color = 'gray',\n",
    "        edgecolor = 'black', label = gene_name2.upper())\n",
    "ax4.set_ylabel('Variant count (log scale)', fontsize = 13)\n",
    "ax4.set_yscale('log')\n",
    "plt.legend(ncol = 2, loc = 'upper center')\n",
    "plt.tight_layout();\n",
    "\n",
    "plt.savefig(r'figures/' + date + '_' + gene_name1.lower() + '_' + gene_name2.lower() +\n",
    "            '_cons_pat.pdf')\n",
    "plt.savefig(r'figures/' + date + '_' + gene_name1.lower() + '_' + gene_name2.lower() +\n",
    "            '_cons_pat.jpg')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Location of as pathogenic characterized variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.subplots(figsize=(12, 8))\n",
    "if gene_name2 != '':\n",
    "    ax1 = plt.subplot(2,1,1)\n",
    "else:\n",
    "    ax1 = plt.subplot(1,1,1)\n",
    "# location of pathogenic variants with frameshift/stop gained gene 1\n",
    "plt.suptitle('Location of pathogenic variants', fontsize=15, fontweight = 'bold', y = 0.95)\n",
    "ax1.set_title(gene_name1.upper() + ' gene', fontsize=12, \n",
    "              fontweight = 'bold', y=1.2, loc = 'left')\n",
    "ax1.scatter(x = df1['POS'][(df1['ClinVar_gnomAD'] == 'Pathogenic') &\n",
    "                           (df1['Consequence_VEP'] == 'frameshift variant')], \n",
    "            y = df1['CADD_PHRED_VEP'][(df1['ClinVar_gnomAD'] == 'Pathogenic') &\n",
    "                                       (df1['Consequence_VEP'] == 'frameshift variant')], \n",
    "            alpha = 0.5, label = 'frameshift variant')\n",
    "ax1.scatter(x = df1['POS'][(df1['ClinVar_gnomAD'] == 'Pathogenic') &\n",
    "                           (df1['Consequence_VEP'] == 'stop gained')], \n",
    "            y = df1['CADD_PHRED_VEP'][(df1['ClinVar_gnomAD'] == 'Pathogenic') &\n",
    "                                      (df1['Consequence_VEP'] == 'stop gained')], \n",
    "            alpha = 0.5, label = 'stop gained')\n",
    "ax1.scatter(x = df1['POS'][(df1['ClinVar_gnomAD'] == 'Pathogenic') &\n",
    "                           (df1['Consequence_VEP'] == 'splice acceptor variant')], \n",
    "            y = df1['CADD_PHRED_VEP'][(df1['ClinVar_gnomAD'] == 'Pathogenic') &\n",
    "                                      (df1['Consequence_VEP'] == 'splice acceptor variant')], \n",
    "            alpha = 0.5, label = 'splice acceptor variant')\n",
    "ax1.scatter(x = df1['POS'][(df1['ClinVar_gnomAD'] == 'Pathogenic') &\n",
    "                           (df1['Consequence_VEP'] == 'splice donor variant')], \n",
    "            y = df1['CADD_PHRED_VEP'][(df1['ClinVar_gnomAD'] == 'Pathogenic') &\n",
    "                                      (df1['Consequence_VEP'] == 'splice donor variant')], \n",
    "            alpha = 0.5, label = 'splice donor variant')\n",
    "plt.legend(bbox_to_anchor = (0.07, 1.4), loc = 'lower left', ncol = 5, fontsize = 12)\n",
    "\n",
    "ax1.set_xlabel('Nucleic position at ' + gene_name1.upper(), fontsize = 12)\n",
    "ax1.set_ylabel('CADD phred score (VEP)', fontsize = 12)\n",
    "ax1.ticklabel_format(style = 'plain')\n",
    "ax1.get_xaxis().set_major_formatter(tick.StrMethodFormatter('{x:,.0f}'))\n",
    "#plt.legend(bbox_to_anchor = (1, 1.12), ncol = 2)\n",
    "\n",
    "if gene_name2 != '':\n",
    "    # consequences of pathogenic variants gene 2\n",
    "    ax2 = plt.subplot(2,1,2, sharey = ax1)\n",
    "    ax2.set_title(gene_name2.upper() + ' gene', fontsize=12, \n",
    "                  fontweight = 'bold', y=1.2, loc = 'left')\n",
    "    ax2.scatter(x = df2['POS'][(df2['ClinVar_gnomAD'] == 'Pathogenic') &\n",
    "                               (df2['Consequence_VEP'] == 'frameshift variant')], \n",
    "                y = df2['CADD_PHRED_VEP'][(df2['ClinVar_gnomAD'] == 'Pathogenic') &\n",
    "                                           (df2['Consequence_VEP'] == 'frameshift variant')], \n",
    "                alpha = 0.5, label = 'frameshift variant')\n",
    "    ax2.scatter(x = df2['POS'][(df2['ClinVar_gnomAD'] == 'Pathogenic') &\n",
    "                               (df2['Consequence_VEP'] == 'stop gained')], \n",
    "                y = df2['CADD_PHRED_VEP'][(df2['ClinVar_gnomAD'] == 'Pathogenic') &\n",
    "                                          (df2['Consequence_VEP'] == 'stop gained')], \n",
    "                alpha = 0.5, label = 'stop gained')\n",
    "    ax2.scatter(x = df2['POS'][(df2['ClinVar_gnomAD'] == 'Pathogenic') &\n",
    "                               (df2['Consequence_VEP'] == 'splice acceptor variant')], \n",
    "                y = df2['CADD_PHRED_VEP'][(df2['ClinVar_gnomAD'] == 'Pathogenic') &\n",
    "                                          (df2['Consequence_VEP'] == 'splice acceptor variant')], \n",
    "                alpha = 0.5, label = 'splice acceptor variant')\n",
    "    ax2.scatter(x = df2['POS'][(df2['ClinVar_gnomAD'] == 'Pathogenic') &\n",
    "                               (df2['Consequence_VEP'] == 'splice donor variant')], \n",
    "                y = df2['CADD_PHRED_VEP'][(df2['ClinVar_gnomAD'] == 'Pathogenic') &\n",
    "                                          (df2['Consequence_VEP'] == 'splice donor variant')], \n",
    "                alpha = 0.5, label = 'splice donor variant')\n",
    "    ax2.set_xlabel('Nucleic position at ' + gene_name2.upper(), fontsize = 12)\n",
    "    ax2.set_ylabel('CADD phred score (VEP)', fontsize = 12)\n",
    "    ax2.ticklabel_format(style = 'plain')\n",
    "    ax2.get_xaxis().set_major_formatter(tick.StrMethodFormatter('{x:,.0f}'))\n",
    "    handles, labels = plt.gca().get_legend_handles_labels()\n",
    "    patch = pplt.Patch(color='grey', label='Exons')\n",
    "    handles.extend([patch])\n",
    "    (y_min1, y_max1) = plt.ylim()\n",
    "    if df1.iloc[0]['Strand'] == '+':\n",
    "        for tup,l in zip(tuple_exons1, range(len(tuple_exons1))):\n",
    "            ax1.add_patch(pplt.Rectangle((int(tup[0]), -.3), abs(int(tup[0])-int(tup[1])), 55, \n",
    "                                         fc = 'black', ec = 'black', alpha = 0.1))\n",
    "            if (l+2)%2 == 0:\n",
    "                ax1.text(int(tup[0])-500, y_max1+1, str((int(l)+2)), size = 10)\n",
    "            else:\n",
    "                ax1.text(int(tup[0])-500, y_max1+3, str((int(l)+2)), size = 10)\n",
    "    elif df1.iloc[0]['Strand'] == '-':\n",
    "        for tup,l in zip(tuple_exons1, range(len(tuple_exons1))):\n",
    "            ax1.add_patch(pplt.Rectangle((int(tup[1]), -.3), abs(int(tup[0])-int(tup[1])), 55, \n",
    "                                         fc = 'black', ec = 'black', alpha = 0.1))\n",
    "            if (l+2)%2 == 0:\n",
    "                ax1.text(int(tup[0])-500, y_max1+1, str((int(l)+2)), size = 12)\n",
    "            else:\n",
    "                ax1.text(int(tup[0])-500, y_max1+5, str((int(l)+2)), size = 12)\n",
    "    if df2.iloc[0]['Strand'] == '+':\n",
    "        for tup,l in zip(tuple_exons2, range(len(tuple_exons2))):\n",
    "            ax2.add_patch(pplt.Rectangle((int(tup[0]), -.3), abs(int(tup[0])-int(tup[1])), 55, fc = 'black', \n",
    "                                         ec = 'black', alpha = 0.1))\n",
    "            if (l+2)%2 == 0:\n",
    "                ax2.text(int(tup[0])-500, y_max1+1, str((int(l)+2)), size = 12)\n",
    "            else:\n",
    "                ax2.text(int(tup[0])-500, y_max1+5, str((int(l)+2)), size = 12)\n",
    "    elif df2.iloc[0]['Strand'] == '-':\n",
    "        for tup,l in zip(tuple_exons2, range(len(tuple_exons2))):\n",
    "            ax2.add_patch(pplt.Rectangle((int(tup[1]), -.3), abs(int(tup[0])-int(tup[1])), 55, fc = 'black', \n",
    "                                         ec = 'black', alpha = 0.1))\n",
    "            if (l+2)%2 == 0:\n",
    "                ax2.text(int(tup[0])-500, y_max1+1, str((int(l)+2)), size = 12)\n",
    "            else:\n",
    "                ax2.text(int(tup[0])-500, y_max1+5, str((int(l)+2)), size = 12)\n",
    "    plt.tight_layout();\n",
    "    plt.savefig(r'figures/' + date + '_' + gene_name1.lower() + '_' + gene_name2.lower() + '_loc_pat.pdf')\n",
    "    plt.savefig(r'figures/' + date + '_' + gene_name1.lower() + '_' + gene_name2.lower() + '_loc_pat.jpg')\n",
    "else:\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(r'figures/' + date + '_' + gene_name1.lower() + '_loc_pat.pdf')\n",
    "    plt.savefig(r'figures/' + date + '_' + gene_name1.lower() + '_loc_pat.jpg')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Deleteriousness of variants"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GnomAD Allele Frequency vs. CADD phred score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df1_af_cadd1 = df1[(df1['AF_gnomAD'].notnull())&(df1['CADD_PHRED_VEP'].notnull())].copy()\n",
    "df1_af_cadd1['AF_to_CADD'] = df1_af_cadd1['AF_gnomAD']/df1_af_cadd1['CADD_PHRED_VEP']\n",
    "p1_l = np.percentile(df1_af_cadd1['AF_to_CADD'], 10)\n",
    "p1_h = np.percentile(df1_af_cadd1['AF_to_CADD'], 90)\n",
    "if gene_name2 != '':\n",
    "    df2_af_cadd2 = df2[(df2['AF_gnomAD'].notnull())&(df2['CADD_PHRED_VEP'].notnull())].copy()\n",
    "    df2_af_cadd2['AF_to_CADD'] = df2_af_cadd2['AF_gnomAD']/df2_af_cadd2['CADD_PHRED_VEP']\n",
    "    p2_l = np.percentile(df1_af_cadd1['AF_to_CADD'], 5)\n",
    "    p2_h = np.percentile(df2_af_cadd2['AF_to_CADD'], 95)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_inall = []\n",
    "df_clin_nct_cbp = []\n",
    "df_nct_cbp = []\n",
    "for df in dfs:\n",
    "    df_inall.append(df[df['ClinVar_gnomAD'].notnull() & \n",
    "                   df['NCT_SAMPLE_NAME'].notnull() & \n",
    "                   df['Annotation_cBP'].notnull() & \n",
    "                   df['Trans_Version_Flossies'].notnull()]) #1:21, 2:17\n",
    "    df_clin_nct_cbp.append(df[df['ClinVar_gnomAD'].notnull() & \n",
    "                   df['NCT_SAMPLE_NAME'].notnull() & \n",
    "                   df['Annotation_cBP'].notnull()]) #1:31, 2:30\n",
    "    df_nct_cbp.append(df[df['NCT_SAMPLE_NAME'].notnull() & \n",
    "                  df['Annotation_cBP'].notnull()]) #1:38, 2:35\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# dfs sorted by clinvar_order\n",
    "afc1 = df1[df1['ClinVar_gnomAD'].notnull()].copy()\n",
    "afc1['ClinVar_gnomAD'] = afc1['ClinVar_gnomAD'].astype(clinvar_order)\n",
    "afc1 = afc1.sort_values('ClinVar_gnomAD')\n",
    "\n",
    "if gene_name2 != '':\n",
    "    afc2 = df2[df2['ClinVar_gnomAD'].notnull()].copy()\n",
    "    afc2['ClinVar_gnomAD'] = afc2['ClinVar_gnomAD'].astype(clinvar_order)\n",
    "    afc2 = afc2.sort_values('ClinVar_gnomAD')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# allele frequency (gnomAD) and cadd score of variants\n",
    "leg_els = [lines.Line2D([], [], color = 'firebrick', marker = 'o', linestyle = '', label = 'Pathogenic'), \n",
    "           lines.Line2D([], [], color = 'red', marker = 'o', linestyle = '', \n",
    "                        label = 'Pathogenic/Likely pathogenic'), \n",
    "           lines.Line2D([], [], color = 'orange', marker = 'o', linestyle = '', label = 'Likely pathogenic'), \n",
    "           lines.Line2D([], [], color = 'yellow', marker = 'o', linestyle = '', \n",
    "                        label = 'Conflicting interpretations of pathogenicity'), \n",
    "           lines.Line2D([], [], color = 'greenyellow', marker = 'o', linestyle = '', label = 'Likely benign'), \n",
    "           lines.Line2D([], [], color = 'limegreen', marker = 'o', linestyle = '', \n",
    "                        label = 'Benign/Likely benign'), \n",
    "           lines.Line2D([], [], color = 'green', marker = 'o', linestyle = '', label = 'Benign')]\n",
    "\n",
    "fig = plt.subplots(figsize = (12,5))\n",
    "plt.suptitle('Allele Frequency (gnomAD) to CADD phred score (VEP)', fontsize = 15, fontweight = 'bold')\n",
    "ax1 = plt.subplot(1,2,1)\n",
    "ax1.set_title(label = gene_name1.upper(), fontsize = 12, fontweight = 'bold')\n",
    "sns.scatterplot(ax = ax1, data = afc1[(afc1['ClinVar_gnomAD'] != 'Uncertain significance')&\n",
    "                                      (afc1['ClinVar_gnomAD'] != 'not provided')], \n",
    "                x = 'CADD_PHRED_VEP', y = 'AF_gnomAD', hue = 'ClinVar_gnomAD', \n",
    "                palette = ['firebrick', 'red', 'orange', 'yellow', 'greenyellow', 'limegreen', 'green', \n",
    "                           'lightgray', 'grey'], alpha = 0.6, legend = False)\n",
    "ax1.set_yscale('log')\n",
    "ax1.set_ylabel('Allele Frequency by gnomAD (log scale)', fontsize = 12)\n",
    "ax1.set_xlabel('CADD phred score (VEP)', fontsize = 12)\n",
    "ax1.tick_params(labelsize = 12)\n",
    "plt.tight_layout();\n",
    "\n",
    "ax2 = plt.subplot(1,2,2, sharex = ax1, sharey = ax1)\n",
    "ax2.set_title(label = gene_name2.upper(), fontsize = 12, fontweight = 'bold')\n",
    "sns.scatterplot(ax = ax2, data = afc2[(afc2['ClinVar_gnomAD'] != 'Uncertain significance')&\n",
    "                                      (afc2['ClinVar_gnomAD'] != 'not provided')], \n",
    "                x = 'CADD_PHRED_VEP', y = 'AF_gnomAD', hue = 'ClinVar_gnomAD', \n",
    "                palette = ['firebrick', 'red', 'orange', 'yellow', 'greenyellow', 'limegreen', 'green', \n",
    "                           'lightgray', 'grey'], alpha = 0.6, legend = False)\n",
    "ax2.set_yscale('log')\n",
    "ax2.set_ylabel('Allele Frequency by gnomAD (log scale)', fontsize = 12)\n",
    "ax2.set_xlabel('CADD phred score (VEP)', fontsize = 12)\n",
    "ax2.tick_params(labelsize = 12)\n",
    "ax2.get_yaxis().set_visible(False)\n",
    "plt.legend(handles = leg_els, bbox_to_anchor = (1.01, 1.02), loc = 'upper left')\n",
    "plt.tight_layout();\n",
    "\n",
    "if len(gene_names) > 1:\n",
    "    plt.savefig(r'figures/' + date + '_' + gene_names[0] + '_' + gene_names[1] + '_cadd_af.pdf')\n",
    "    plt.savefig(r'figures/' + date + '_' + gene_names[0] + '_' + gene_names[1] + '_cadd_af.jpg')\n",
    "else:\n",
    "    plt.savefig(r'figures/' + date + '_' + gene_names[0] + '_cadd_af.pdf')\n",
    "    plt.savefig(r'figures/' + date + '_' + gene_names[0] + '_cadd_af.jpg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# allele frequency and cadd score of some variants (CADD > 15, AF > 0.001)\n",
    "fig = plt.subplots(figsize = (12,5))\n",
    "plt.suptitle('Allele Frequency (gnomAD) to CADD phred score (VEP) of conflicting variants', \n",
    "             fontsize = 15, fontweight = 'bold')\n",
    "ax1 = plt.subplot(1,2,1)\n",
    "ax1.set_title(gene_name1.upper(), fontweight = 'bold')\n",
    "sns.scatterplot(ax = ax1, data = df1[(df1['CADD_PHRED_VEP'] > 15)&(df1['AF_gnomAD'] > 0.001)], \n",
    "                x = 'CADD_PHRED_VEP', y = 'AF_gnomAD', hue = 'ClinVar_gnomAD', legend = False)\n",
    "ax1.set_yscale('log', fontsize = 12)\n",
    "ax1.set_ylabel('Allele Frequency by gnomAD (log scale)', fontsize = 12)\n",
    "ax1.set_xlabel('CADD phred score (VEP)', fontsize = 12)\n",
    "plt.tight_layout()\n",
    "\n",
    "ax2 = plt.subplot(1,2,2, sharex = ax1, sharey = ax1)\n",
    "ax2.set_title(gene_name2.upper(), fontweight = 'bold')\n",
    "sns.scatterplot(ax = ax2, data = df2[(df2['CADD_PHRED_VEP'] > 15)&(df2['AF_gnomAD'] > 0.001)], \n",
    "                x = 'CADD_PHRED_VEP', y = 'AF_gnomAD', hue = 'ClinVar_gnomAD')\n",
    "ax2.\n",
    "ax2.set_yscale('log')\n",
    "ax2.set_ylabel('Allele Frequency by gnomAD (log scale)')\n",
    "ax2.set_xlabel('CADD phred score (VEP)')\n",
    "#ax2.legend(bbox_to_anchor = (-0.1, 1))\n",
    "plt.tight_layout();\n",
    "\n",
    "if len(gene_names) > 1:\n",
    "    plt.savefig(r'figures/' + date + '_' + gene_names[0] + '_' + gene_names[1] + '_cadd_af_confl.pdf')\n",
    "    plt.savefig(r'figures/' + date + '_' + gene_names[0] + '_' + gene_names[1] + '_cadd_af_confl.jpg')\n",
    "else:\n",
    "    plt.savefig(r'figures/' + date + '_' + gene_names[0] + '_cadd_af_confl.pdf')\n",
    "    plt.savefig(r'figures/' + date + '_' + gene_names[0] + '_cadd_af_confl.jpg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# allele frequency of Flossies and cadd score of variants\n",
    "leg_els = [lines.Line2D([], [], color = 'firebrick', marker = 'o', linestyle = '', label = 'Pathogenic'), \n",
    "           lines.Line2D([], [], color = 'red', marker = 'o', linestyle = '', \n",
    "                        label = 'Pathogenic/Likely pathogenic'), \n",
    "           lines.Line2D([], [], color = 'orange', marker = 'o', linestyle = '', label = 'Likely pathogenic'), \n",
    "           lines.Line2D([], [], color = 'yellow', marker = 'o', linestyle = '', \n",
    "                        label = 'Conflicting interpretations of pathogenicity'), \n",
    "           lines.Line2D([], [], color = 'greenyellow', marker = 'o', linestyle = '', label = 'Likely benign'), \n",
    "           lines.Line2D([], [], color = 'limegreen', marker = 'o', linestyle = '', \n",
    "                        label = 'Benign/Likely benign'), \n",
    "           lines.Line2D([], [], color = 'green', marker = 'o', linestyle = '', label = 'Benign')]\n",
    "\n",
    "fig = plt.subplots(figsize = (12,5))\n",
    "plt.suptitle('Allele Frequency (Flossies) to CADD phred score (VEP)', fontsize = 15, fontweight = 'bold')\n",
    "ax1 = plt.subplot(1,2,1)\n",
    "ax1.set_title(label = gene_name1.upper(), fontsize = 12, fontweight = 'bold')\n",
    "sns.scatterplot(ax = ax1, data = afc1[(afc1['ClinVar_gnomAD'] != 'Uncertain significance')&\n",
    "                                      (afc1['ClinVar_gnomAD'] != 'not provided')], \n",
    "                x = 'CADD_PHRED_VEP', y = 'Overall_Frequency_Flossies', hue = 'ClinVar_gnomAD', \n",
    "                palette = ['firebrick', 'red', 'orange', 'yellow', 'greenyellow', 'limegreen', 'green', \n",
    "                           'lightgray', 'grey'], alpha = 0.6, legend = False)\n",
    "ax1.set_yscale('log')\n",
    "ax1.set_ylabel('Allele Frequency by Flossies (log scale)', fontsize = 12)\n",
    "ax1.set_xlabel('CADD phred score (VEP)', fontsize = 12)\n",
    "ax1.tick_params(labelsize = 12)\n",
    "plt.tight_layout();\n",
    "\n",
    "ax2 = plt.subplot(1,2,2, sharex = ax1, sharey = ax1)\n",
    "ax2.set_title(label = gene_name2.upper(), fontsize = 12, fontweight = 'bold')\n",
    "sns.scatterplot(ax = ax2, data = afc2[(afc2['ClinVar_gnomAD'] != 'Uncertain significance')&\n",
    "                                      (afc2['ClinVar_gnomAD'] != 'not provided')], \n",
    "                x = 'CADD_PHRED_VEP', y = 'Overall_Frequency_Flossies', hue = 'ClinVar_gnomAD', \n",
    "                palette = ['firebrick', 'red', 'orange', 'yellow', 'greenyellow', 'limegreen', 'green', \n",
    "                           'lightgray', 'grey'], alpha = 0.6, legend = False)\n",
    "ax2.set_yscale('log')\n",
    "ax2.set_ylabel('Allele Frequency by Flossies (log scale)', fontsize = 12)\n",
    "ax2.set_xlabel('CADD phred score (VEP)', fontsize = 12)\n",
    "ax2.tick_params(labelsize = 12)\n",
    "ax2.get_yaxis().set_visible(False)\n",
    "plt.legend(handles = leg_els, bbox_to_anchor = (1.01, 1.02), loc = 'upper left')\n",
    "plt.tight_layout()\n",
    "\n",
    "if len(gene_names) > 1:\n",
    "    plt.savefig(r'figures/' + date + '_' + gene_names[0] + '_' + gene_names[1] + '_cadd_flos.pdf')\n",
    "    plt.savefig(r'figures/' + date + '_' + gene_names[0] + '_' + gene_names[1] + '_cadd_flos.jpg')\n",
    "else:\n",
    "    plt.savefig(r'figures/' + date + '_' + gene_names[0] + '_cadd_flos.pdf')\n",
    "    plt.savefig(r'figures/' + date + '_' + gene_names[0] + '_cadd_flos.jpg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# allele frequency and cadd score of some variants (CADD > 15, AF > 0.001)\n",
    "fig = plt.subplots(figsize = (12,5))\n",
    "plt.suptitle('Allele Frequency (Flossies) to CADD phred score (VEP) of conflicting variants', \n",
    "             fontsize = 15, fontweight = 'bold')\n",
    "ax1 = plt.subplot(1,2,1)\n",
    "ax1.set_title(gene_name1.upper(), fontweight = 'bold')\n",
    "sns.scatterplot(ax = ax1, data = df1[(df1['CADD_PHRED_VEP'] > 15)&(df1['Overall_Frequency_Flossies'] > 0.001)],\n",
    "                x = 'CADD_PHRED_VEP', y = 'AF_gnomAD', hue = 'ClinVar_gnomAD')\n",
    "ax1.set_yscale('log')\n",
    "ax1.set_ylabel('Allele Frequency by Flossies (log scale)')\n",
    "ax1.set_xlabel('CADD phred score (VEP)')\n",
    "plt.tight_layout()\n",
    "\n",
    "ax2 = plt.subplot(1,2,2, sharex = ax1, sharey = ax1)\n",
    "ax2.set_title(gene_name2.upper(), fontweight = 'bold')\n",
    "sns.scatterplot(ax = ax2, data = df2[(df2['CADD_PHRED_VEP'] > 15)&(df2['Overall_Frequency_Flossies'] > 0.001)],\n",
    "                x = 'CADD_PHRED_VEP', y = 'AF_gnomAD', hue = 'ClinVar_gnomAD')\n",
    "ax2.set_yscale('log')\n",
    "ax2.set_ylabel('Allele Frequency by FLossies (log scale)')\n",
    "ax2.set_xlabel('CADD phred score (VEP)')\n",
    "#ax2.legend(bbox_to_anchor = (-0.1, 0.8))\n",
    "plt.tight_layout()\n",
    "\n",
    ";\n",
    "if len(gene_names) > 1:\n",
    "    plt.savefig(r'figures/' + date + '_' + gene_names[0] + '_' + gene_names[1] + '_cadd_floss_confl.pdf')\n",
    "    plt.savefig(r'figures/' + date + '_' + gene_names[0] + '_' + gene_names[1] + '_cadd_floss_confl.jpg')\n",
    "else:\n",
    "    plt.savefig(r'figures/' + date + '_' + gene_names[0] + '_cadd_floss_confl.pdf')\n",
    "    plt.savefig(r'figures/' + date + '_' + gene_names[0] + '_cadd_floss_confl.jpg')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Division in groups for prediction of deleteriousness"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "divide variants into groups of deleteriousness\n",
    "\n",
    "set thresholds for deleteriosness of variants\n",
    "- whole df: 42623 variants\n",
    "- AF_gnomAD < 0.001 (allele frequency, 3600 variants of 3670)\n",
    "- CADD_PHRED_VEP > 15 (rank of variant to substitution in whole human reference genome, 15571 variants of 42348)\n",
    "- Ada_score_VEP > 0.6 (splicing effect score, 710 variants of 1884)\n",
    "- SIFT_VEP_num < 0.05 (resulting changed protein function by aa change, 10553 of 22772)\n",
    "- PolyPhen_VEP_num > 0.6 (resulting changed protein structure/function by aa change, 6073 of 22772)\n",
    "- Impact_VEP == 'HIGH'\n",
    "- ClinVar_gnomAD == 'Pathogenic', 'Pathogenic/Likely pathogenic', 'Conflicting interpretations of pathogenicity', 'not provided', 'Uncertain significance'\n",
    "\n",
    "results:\n",
    "- for everything: no result\n",
    "- if no Ada_score (for splice effects) --> one result: start_lost (of 9x 'start lost')\n",
    "- if add. no Impact_VEP --> 366 results for 'cplt_path_group'\n",
    "\n",
    "\n",
    "'''\n",
    "\n",
    "# variants completely categorized as deleterious\n",
    "cplt_path_group = []\n",
    "cplt_ben_group = []\n",
    "for df in dfs:\n",
    "    cplt_path_group.append(df[(df['AF_gnomAD'] < 0.001) &\n",
    "                              (df['SIFT_VEP_num'] < 0.05) &\n",
    "                              (df['CADD_PHRED_VEP'] > 15) &\n",
    "                              (df['PolyPhen_VEP_num'] > 0.6) &\n",
    "                              (df['ClinVar_gnomAD'].isin(['Pathogenic', 'Pathogenic/Likely pathogenic', \n",
    "                                                          'Conflicting interpretations of pathogenicity', \n",
    "                                                          'not provided', 'Uncertain significance']))])\n",
    "    cplt_ben_group.append(df[(df['AF_gnomAD'] >= 0.001) &\n",
    "                             (df['SIFT_VEP_num'] >= 0.05) &\n",
    "                             (df['CADD_PHRED_VEP'] <= 15) &\n",
    "                             (df['PolyPhen_VEP_num'] <= 0.6)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# SIFT_VEP vs. PolyPhen_VEP\n",
    "fig = plt.subplots(figsize = (10,10))\n",
    "\n",
    "#for df, g in zip(dfs, gene_names):\n",
    "ax1 = plt.subplot(2,1,1)\n",
    "ax1.set_title('SIFT score (VEP) to PolyPhen score (VEP) in ' + gene_name1.upper() + ' gene', \n",
    "              fontsize = 15, fontweight = 'bold')\n",
    "ax1 = sns.kdeplot(data = df1, x = 'SIFT_VEP_num', y = 'PolyPhen_VEP_num', fill = True, thresh = 0, \n",
    "                  levels = 1000, cmap = 'Blues', cbar = True)\n",
    "ax1.set_ylabel('SIFT score by VEP', fontsize = 12)\n",
    "ax1.set_xlabel('PolyPhen score by VEP', fontsize = 12)\n",
    "#plt.legend()\n",
    "\n",
    "ax2 = plt.subplot(2,1,2)\n",
    "ax2.set_title('SIFT score (VEP) to PolyPhen score (VEP) in ' + gene_name2.upper() + ' gene', \n",
    "              fontsize = 15, fontweight = 'bold')\n",
    "ax2 = sns.kdeplot(data = df2, x = 'SIFT_VEP_num', y = 'PolyPhen_VEP_num', fill = True, thresh = 0, \n",
    "                  levels = 1000, cmap = 'Blues', cbar = True)\n",
    "ax2.set_ylabel('SIFT score by VEP', fontsize = 12)\n",
    "ax2.set_xlabel('PolyPhen score by VEP', fontsize = 12)\n",
    "plt.tight_layout()\n",
    "\n",
    "if len(gene_names) > 1:\n",
    "    plt.savefig(r'figures/' + date + '_' + gene_names[0] + '_' + gene_names[1] + '_sift_pp_dens.pdf')\n",
    "    plt.savefig(r'figures/' + date + '_' + gene_names[0] + '_' + gene_names[1] + '_sift_pp_dens.jpg')\n",
    "else:\n",
    "    plt.savefig(r'figures/' + date + '_' + gene_names[0] + '_sift_pp_dens.pdf')\n",
    "    plt.savefig(r'figures/' + date + '_' + gene_names[0] + '_sift_pp_dens.jpg')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Stop gain"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Where are variants with gained stop codons?\n",
    "Which variants are how annotated in ClinVar, which in cBioPortal?\n",
    "'''\n",
    "\n",
    "# for stop gain --> where are variants with gained stop codons, how annotated in ClinVar, which in cBioPortal\n",
    "fig = plt.subplots(figsize = (11,8))\n",
    "plt.suptitle('Stop gained variants  with ClinVar annotations', fontweight = 'bold', fontsize = 15, \n",
    "             y = 0.97, x = 0.5)\n",
    "ax1 = plt.subplot(2,1,1)\n",
    "sns.scatterplot(ax = ax1, x = df1['POS'][df1['Consequence_VEP'] == 'stop gained'], \n",
    "                y = df1['CADD_PHRED_VEP'][df1['Consequence_VEP'] == 'stop gained'], \n",
    "                hue = df1['ClinVar_gnomAD'][df1['Consequence_VEP'] == 'stop gained'].fillna('Not in GnomAD'), \n",
    "                edgecolor = None, alpha = 0.5, legend = False)\n",
    "ax1.set_title(gene_name1.upper(), fontsize=13, fontweight = 'bold', loc = 'left')\n",
    "ax1.set_xlabel('Nucleic position at ' + gene_name1.upper(), fontsize = 12)\n",
    "ax1.set_ylabel('CADD phred score (by VEP)', fontsize = 12)\n",
    "ax1.ticklabel_format(style = 'plain')\n",
    "ax1.get_xaxis().set_major_formatter(tick.StrMethodFormatter('{x:,.0f}'));\n",
    "plt.tight_layout()\n",
    "\n",
    "ax2 = plt.subplot(2,1,2)\n",
    "sns.scatterplot(ax = ax2, x = df2['POS'][df2['Consequence_VEP'] == 'stop gained'], \n",
    "                y = df2['CADD_PHRED_VEP'][df2['Consequence_VEP'] == 'stop gained'], \n",
    "                hue = df2['ClinVar_gnomAD'][df2['Consequence_VEP'] == 'stop gained'].fillna('Not in GnomAD'), \n",
    "                edgecolor = None, alpha = 0.5)\n",
    "ax2.set_title(gene_name2.upper(), loc = 'left', fontsize=13, fontweight = 'bold')\n",
    "ax2.set_xlabel('Nucleic position at ' + gene_name2.upper(), fontsize = 12)\n",
    "ax2.set_ylabel('CADD phred score (by VEP)', fontsize = 12)\n",
    "ax2.ticklabel_format(style = 'plain')\n",
    "ax2.get_xaxis().set_major_formatter(tick.StrMethodFormatter('{x:,.0f}'))\n",
    "ax2.legend(bbox_to_anchor = (0, -0.2), loc = 'upper left', borderaxespad = 0, ncol = 5)\n",
    "plt.tight_layout\n",
    "plt.gcf().subplots_adjust(bottom = 0.12, top = 0.9, left = 0.05, right = 0.95);\n",
    "plt.savefig(r'figures/' + date + '_' + gene_name2.lower() + '_' + \n",
    "            gene_name1.lower() + '_stop_gains.pdf')\n",
    "plt.savefig(r'figures/' + date + '_' + gene_name2.lower() + '_' + \n",
    "            gene_name1.lower() + '_stop_gains.jpeg')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
